Mathlib/RingTheory/Bialgebra/TensorProduct.lean

```diff
@@ -22,128 +22,129 @@ namespace Bialgebra.TensorProduct
-  { Coalgebra.TensorProduct.lid R A, Algebra.TensorProduct.lid R A with }
+protected noncomputable def lid : R ⊗[R] B ≃ₐc[R] B :=
+  { Coalgebra.TensorProduct.lid R B, Algebra.TensorProduct.lid R B with }
 
 @[simp]
 theorem lid_toCoalgEquiv :
-    (Bialgebra.TensorProduct.lid R A : R ⊗[R] A ≃ₗc[R] A) = Coalgebra.TensorProduct.lid R A := rfl
+    (Bialgebra.TensorProduct.lid R B : R ⊗[R] B ≃ₗc[R] B) = Coalgebra.TensorProduct.lid R B := rfl
 
 @[simp]
 theorem lid_toAlgEquiv :
-    (Bialgebra.TensorProduct.lid R A : R ⊗[R] A ≃ₐ[R] A) = Algebra.TensorProduct.lid R A := rfl
+    (Bialgebra.TensorProduct.lid R B : R ⊗[R] B ≃ₐ[R] B) = Algebra.TensorProduct.lid R B := rfl
 
 @[simp]
-theorem lid_tmul (r : R) (a : A) : Bialgebra.TensorProduct.lid R A (r ⊗ₜ a) = r • a := rfl
+theorem lid_tmul (r : R) (a : B) : Bialgebra.TensorProduct.lid R B (r ⊗ₜ a) = r • a := rfl
 
 @[simp]
-theorem lid_symm_apply (a : A) : (Bialgebra.TensorProduct.lid R A).symm a = 1 ⊗ₜ a := rfl
+theorem lid_symm_apply (a : B) : (Bialgebra.TensorProduct.lid R B).symm a = 1 ⊗ₜ a := rfl
 
-/- TODO: make this defeq, which would involve adding a heterobasic version of
-`Coalgebra.TensorProduct.rid`. -/
 theorem coalgebra_rid_eq_algebra_rid_apply (x : A ⊗[R] R) :
-    Coalgebra.TensorProduct.rid R A x = Algebra.TensorProduct.rid R R A x := rfl
+    Coalgebra.TensorProduct.rid R S A x = Algebra.TensorProduct.rid R R A x := rfl
 
-variable (R A) in
+variable (R S A) in
 /-- The base ring is a right identity for the tensor product of bialgebras, up to
 bialgebra equivalence. -/
-protected noncomputable def rid : A ⊗[R] R ≃ₐc[R] A where
-  toCoalgEquiv := Coalgebra.TensorProduct.rid R A
+protected noncomputable def rid : A ⊗[R] R ≃ₐc[S] A where
+  toCoalgEquiv := Coalgebra.TensorProduct.rid R S A
   map_mul' x y := by
     simp only [CoalgEquiv.toCoalgHom_eq_coe, CoalgHom.toLinearMap_eq_coe, AddHom.toFun_eq_coe,
       LinearMap.coe_toAddHom, CoalgHom.coe_toLinearMap, CoalgHom.coe_coe,
       coalgebra_rid_eq_algebra_rid_apply, map_mul]
 
 @[simp]
 theorem rid_toCoalgEquiv :
-    (TensorProduct.rid R A : A ⊗[R] R ≃ₗc[R] A) = Coalgebra.TensorProduct.rid R A := rfl
+    (TensorProduct.rid R S A : A ⊗[R] R ≃ₗc[S] A) = Coalgebra.TensorProduct.rid R S A := rfl
 
 @[simp]
 theorem rid_toAlgEquiv :
-    (Bialgebra.TensorProduct.rid R A : A ⊗[R] R ≃ₐ[R] A) = Algebra.TensorProduct.rid R R A := by
+    (Bialgebra.TensorProduct.rid R S A : A ⊗[R] R ≃ₐ[S] A) = Algebra.TensorProduct.rid R S A := by
   ext x
   exact coalgebra_rid_eq_algebra_rid_apply x
 
 @[simp]
-theorem rid_tmul (r : R) (a : A) : Bialgebra.TensorProduct.rid R A (a ⊗ₜ r) = r • a := rfl
+theorem rid_tmul (r : R) (a : A) : Bialgebra.TensorProduct.rid R S A (a ⊗ₜ r) = r • a := rfl
 
 @[simp]
-theorem rid_symm_apply (a : A) : (Bialgebra.TensorProduct.rid R A).symm a = a ⊗ₜ 1 := rfl
+theorem rid_symm_apply (a : A) : (Bialgebra.TensorProduct.rid R S A).symm a = a ⊗ₜ 1 := rfl
 
 end Bialgebra.TensorProduct
 namespace BialgHom
```

Comment:
Do you think this section should be heterogenized too?

Reply:
I think `ltensor` and `tensor` are over fixed base for linear maps too and if you want general base you should just `Bialgebra.TensorProduct.map` manually.

Reply:
Can you check that holds today for CoAlgHom and AlgHom?

Reply:
We do not have `AlgHom.lTensor`, and `CoalgHom.lTensor` is currently over fixed base purely because everything for `CoalgHom` is over a fixed base before this PR. I didn't touch them in this PR too.

Reply:
Sure, lets leave that for when we need it then.